controller/linodecluster_controller.go

```diff
@@ -18,45 +18,220 @@ package controller
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.16.0/pkg/reconcile
+
 func (r *LinodeClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultedLoopTimeout(r.ReconcileTimeout))
+	defer cancel()
+
+	logger := ctrl.LoggerFrom(ctx).WithName("LinodeClusterReconciler").WithValues("name", req.NamespacedName.String())
+	linodeCluster := &infrav1alpha1.LinodeCluster{}
+	if err := r.Client.Get(ctx, req.NamespacedName, linodeCluster); err != nil {
+		logger.Info("Failed to fetch Linode cluster", "error", err.Error())
+
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	cluster, err := kutil.GetOwnerCluster(ctx, r.Client, linodeCluster.ObjectMeta)
+	if err != nil {
+		logger.Info("Failed to get owner cluster", "error", err.Error())
 
-	// TODO(user): your logic here
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	} else if cluster == nil {
+		logger.Info("Machine Controller has not yet set OwnerRef, skipping reconciliation")
+
+		return ctrl.Result{}, nil
+	}
+	if annotations.IsPaused(cluster, linodeCluster) {
+		logger.Info("LinodeCluster of linked Cluster is marked as paused. Won't reconcile")
+
+		return ctrl.Result{}, nil
+	}
+	// Create the cluster scope.
+	clusterScope, err := scope.NewClusterScope(
+		r.LinodeApiKey,
+		scope.ClusterScopeParams{
+			Client:        r.Client,
+			Cluster:       cluster,
+			LinodeCluster: linodeCluster,
+		})
+	if err != nil {
+		logger.Info("Failed to create cluster scope", "error", err.Error())
+
+		return ctrl.Result{}, fmt.Errorf("failed to create cluster scope: %w", err)
+	}
+
+	return r.reconcile(ctx, clusterScope, logger)
+}
+
+func (r *LinodeClusterReconciler) reconcile(
+	ctx context.Context,
+	clusterScope *scope.ClusterScope,
+	logger logr.Logger,
+) (res ctrl.Result, reterr error) {
+	res = ctrl.Result{}
+
+	clusterScope.LinodeCluster.Status.Ready = false
+	clusterScope.LinodeCluster.Status.FailureReason = nil
+	clusterScope.LinodeCluster.Status.FailureMessage = util.Pointer("")
+
+	// Always close the scope when exiting this function so we can persist any LinodeCluster changes.
+	defer func() {
+		// Filter out any IsNotFound message since client.IgnoreNotFound does not handle aggregate errors
+		if err := clusterScope.Close(); utilerrors.FilterOut(err, apierrors.IsNotFound) != nil && reterr == nil {
+			logger.Error(err, "failed to patch LinodeCluster")
+			reterr = err
+		}
+	}()
+
+	// Handle deleted clusters
+	if !clusterScope.LinodeCluster.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, r.reconcileDelete(ctx, logger, clusterScope)
+	}
+
+	controllerutil.AddFinalizer(clusterScope.LinodeCluster, infrav1alpha1.GroupVersion.String())
```

Comment:
i see some capi providers will call a patch immediately after adding the finalizer to make sure it's there before doing anything else e.g. https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/controllers/awscluster_controller.go#L275-L280 

maybe worth doing too?

Reply:
moved finalizer logic into a function that always does a patch immediately after